File: services/web-backend/src/engine/game-flow.ts
/* ------------------------------------------------------------------ */
/*  Game flow — round lifecycle, scoring, timer management             */
/*  Port of game_flow.py                                               */
/* ------------------------------------------------------------------ */

import type { Room, Player, Difficulty } from './types.js'
import type { AnswerRecord } from './types.js'
import { store } from './room-store.js'
import { calculateScore, isAnswerCorrect } from './scoring.js'
import { notifyBackend } from './notifications.js'
import logger from '../lib/logger.js'

/* ------------------------------------------------------------------ */
/*  Constants                                                          */
/* ------------------------------------------------------------------ */

/** Duration in ms the feedback screen is shown before advancing. */
export const FEEDBACK_DELAY_MS = 5_000

/* ------------------------------------------------------------------ */
/*  Helpers                                                            */
/* ------------------------------------------------------------------ */

function activePlayers(room: Room): Player[] {
  return [...room.players.values()].filter(p => !p.disconnected)
}

function activePlayersCount(room: Room): number {
  return activePlayers(room).length
}

/* ------------------------------------------------------------------ */
/*  Shuffle (Fisher–Yates)                                             */
/* ------------------------------------------------------------------ */

export function shuffleQuestions(room: Room): void {
  const ids = room.questions.map(q => q.id)
  for (let i = ids.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[ids[i], ids[j]] = [ids[j], ids[i]]
  }
  room.shuffledQuestionIds = ids
}

/* ------------------------------------------------------------------ */
/*  Abstract GameFlow                                                  */
/* ------------------------------------------------------------------ */

export abstract class GameFlow {
  /* ── Hooks ─────────────────────────────────────────────────────── */

  abstract shouldFinishRound(room: Room): boolean

  abstract shouldEliminate(player: Player): boolean

  /* ── Callbacks ──────────────────────────────────────────────────── */

  async onAnswer(roomId: string): Promise<void> {
    const room = store.get(roomId)
    if (!room) return
    if (this.shouldFinishRound(room)) {
      await this.finishRound(roomId)
    }
  }

  async onRemovePlayer(roomId: string): Promise<void> {
    const room = store.get(roomId)
    if (!room) return
    if (
      room.status === 'playing' &&
      room.feedbackUntil === null &&
      this.shouldFinishRound(room)
    ) {
      await this.finishRound(roomId)
    }
  }

  /* ── Round lifecycle ───────────────────────────────────────────── */

  /**
   * Starts the game for a room: shuffles questions, sets status,
   * initialises round state, starts the deadline timer.
   *
   * Called from the facade's startGame() and auto-start for solo.
   */
  startGame(roomId: string, _playerId: string | null): void {
    const room = store.getOrThrow(roomId)

    shuffleQuestions(room)
    room.status = 'playing'
    room.currentQuestionIndex = 0
    room.answeredPlayers.clear()
    room.currentRoundAnswers.clear()
    room.feedbackUntil = null

    this.scheduleDeadline(roomId)
  }

  /**
   * Called when the round timer expires (deadline) or when all
   * active players have answered.
   */
  async finishRound(roomId: string): Promise<void> {
    const room = store.get(roomId)
    if (!room || room.status !== 'playing') return

    // Guard: prevent double-finish (race: deadline timer + all answered)
    if (room.feedbackUntil !== null) return

    // Cancel the deadline timer
    if (room.deadlineTimer) {
      clearTimeout(room.deadlineTimer)
      room.deadlineTimer = null
    }

    // Fill unanswered players as timeout
    for (const player of activePlayers(room)) {
      if (!room.currentRoundAnswers.has(player.id)) {
        room.currentRoundAnswers.set(player.id, {
          questionId: room.shuffledQuestionIds[room.currentQuestionIndex] ?? 0,
          selectedChoices: [],
          elapsed: room.timer,
          timeout: true,
        })
        room.answeredPlayers.add(player.id)
      }
    }

    // Score round
    const results = this.scoreRound(room)

    const currentQid = room.shuffledQuestionIds[room.currentQuestionIndex] ?? null
    const qstate = currentQid !== null
      ? room.questions.find(q => q.id === currentQid) ?? null
      : null
    const correctChoices = qstate ? qstate.correctChoices : []

    room.feedbackUntil = Date.now() + FEEDBACK_DELAY_MS

    // Notify backend
    await notifyBackend(roomId, 'question-finished', {
      question_id: currentQid,
      correct_choices: correctChoices,
      results,
    })

    logger.info(
      { event: 'round-finished', roomId, questionId: currentQid, resultCount: results.length },
      'Round finished',
    )

    // Schedule advance after feedback
    this.scheduleAdvance(roomId)
  }

  /**
   * Moves to the next question.  If there are no more questions the
   * game is ended and results are sent.
   */
  async advanceQuestion(roomId: string): Promise<void> {
    const room = store.get(roomId)
    if (!room || room.status !== 'playing') return

    room.currentQuestionIndex++

    // Clear round state
    room.answeredPlayers.clear()
    room.currentRoundAnswers.clear()
    room.feedbackUntil = null

    if (room.advanceTimer) {
      clearTimeout(room.advanceTimer)
      room.advanceTimer = null
    }

    // Check if game is over
    if (room.currentQuestionIndex >= room.shuffledQuestionIds.length) {
      for (const player of room.players.values()) {
        player.finished = true
      }
      room.status = 'finished'

      await this.sendResults(roomId)
      await notifyBackend(roomId, 'game-finished', {})

      logger.info({ event: 'game-finished', roomId }, 'Game finished')
      return
    }

    // Start the next round
    this.scheduleDeadline(roomId)

    await notifyBackend(roomId, 'next-question', {
      question_index: room.currentQuestionIndex,
    })

    logger.info(
      { event: 'next-question', roomId, index: room.currentQuestionIndex },
      'Advancing to next question',
    )
  }

  /* ── Timer management ──────────────────────────────────────────── */

  private scheduleDeadline(roomId: string): void {
    const room = store.get(roomId)
    if (!room) return

    // Clear any existing deadline timer first
    if (room.deadlineTimer) {
      clearTimeout(room.deadlineTimer)
    }

    room.deadlineTimer = setTimeout(() => {
      this.finishRound(roomId).catch(err => {
        logger.error({ err, roomId }, 'Deadline finishRound failed')
      })
    }, room.timer * 1000)
  }

  private scheduleAdvance(roomId: string): void {
    const room = store.get(roomId)
    if (!room) return

    // Clear any existing advance timer first
    if (room.advanceTimer) {
      clearTimeout(room.advanceTimer)
    }

    room.advanceTimer = setTimeout(() => {
      this.advanceQuestion(roomId).catch(err => {
        logger.error({ err, roomId }, 'Advance question failed')
      })
    }, FEEDBACK_DELAY_MS)
  }

  /* ── Scoring helpers ───────────────────────────────────────────── */

  /**
   * Score a completed round.  Mutates player stats on the room.
   * Returns the result entries to send via the 'question-finished'
   * notification.
   */
  protected scoreRound(room: Room): Array<{
    player_id: string
    nickname: string
    correct: boolean
    points: number
    bonus: number
    streak: number
    cumulative_time: number
  }> {
    const currentQid = room.shuffledQuestionIds[room.currentQuestionIndex]
    const qstate = room.questions.find(q => q.id === currentQid)
    if (!qstate) return []

    // Determine which players answered correctly (within time)
    const correctPlayers = new Set<string>()
    for (const [pid, ra] of room.currentRoundAnswers) {
      if (!ra.timeout && isAnswerCorrect(qstate.correctChoices, ra.selectedChoices, qstate.questionType)) {
        correctPlayers.add(pid)
      }
    }

    // Sort answers by elapsed time (fastest first)
    const sortedEntries = [...room.currentRoundAnswers.entries()].sort(
      (a, b) => a[1].elapsed - b[1].elapsed,
    )

    const results: Array<{
      player_id: string
      nickname: string
      correct: boolean
      points: number
      bonus: number
      streak: number
      cumulative_time: number
    }> = []

    for (const [pid, ra] of sortedEntries) {
      const player = room.players.get(pid)
      if (!player) continue

      const oldStreak = player.streak

      let correct: boolean
      let points: number
      let bonus: number
      let newStreak: number
      let timeSpent: number

      if (ra.timeout) {
        correct = false
        points = 0
        bonus = 0
        newStreak = 0
        player.streak = 0
        timeSpent = room.timer
      } else {
        correct = isAnswerCorrect(
          qstate.correctChoices,
          ra.selectedChoices,
          qstate.questionType,
        )
        timeSpent = Math.min(ra.elapsed, room.timer)

        if (correct) {
          const firstCorrect = pid === [...correctPlayers][0]
          const aloneCorrect = correctPlayers.size === 1
          const ctx = {
            mode: room.mode,
            playerCount: activePlayersCount(room),
            difficulty: qstate.difficulty as Difficulty,
            isCorrect: true,
            currentStreak: oldStreak,
            firstCorrect,
            aloneCorrect,
          }
          const sr = calculateScore(ctx)
          points = sr.total
          bonus = sr.bonus
          newStreak = sr.newStreak
          player.streak = newStreak
        } else {
          points = 0
          bonus = 0
          newStreak = oldStreak
          player.streak = 0
        }
      }

      player.score += points
      player.cumulativeTime += timeSpent

      // Record answer for final results
      room.answers.push({
        playerId: pid,
        questionId: qstate.id,
        correct,
        timeSpent,
      })

      results.push({
        player_id: pid,
        nickname: player.nickname,
        correct,
        points,
        bonus,
        streak: correct ? player.streak : 0,
        cumulative_time: player.cumulativeTime,
      })
    }

    return results
  }

  /**
   * Send final results to the backend for persistence.
   */
  protected async sendResults(roomId: string): Promise<void> {
    const room = store.get(roomId)
    if (!room) return

    const sortedPlayers = [...room.players.values()].sort(
      (a, b) => b.score - a.score || a.cumulativeTime - b.cumulativeTime,
    )

    const scores = sortedPlayers.map(p => ({
      player_id: p.id,
      nickname: p.nickname,
      score: p.score,
      streak: p.streak,
      cumulative_time: p.cumulativeTime,
    }))

    const answers = room.answers.map(a => ({
      player_id: a.playerId,
      question_id: a.questionId,
      correct: a.correct,
      time_spent: a.timeSpent,
    }))

    await notifyBackend(roomId, 'results', { scores, answers })
  }
}

/* ------------------------------------------------------------------ */
/*  ClassicFlow (default game mode)                                    */
/* ------------------------------------------------------------------ */

export class ClassicFlow extends GameFlow {
  /** Round finishes when all active players have answered. */
  shouldFinishRound(room: Room): boolean {
    const active = activePlayers(room)
    if (active.length === 0) return false
    return active.every(p => room.answeredPlayers.has(p.id))
  }

  /** Classic mode never eliminates players mid-game. */
  shouldEliminate(_player: Player): boolean {
    return false
  }
}

/** Singleton flow instance. */
export const flow = new ClassicFlow()

File: services/quiz-engine/src/game_flow.py
from __future__ import annotations

import asyncio
import time
from abc import ABC, abstractmethod
from typing import Any

from src.logger import logger
from src.notifications import notify_backend
from src.room_store import AnswerRecord, GameStatus, RoundAnswer, store
from src.scoring import ScoreCalculator, ScoreResult, ScoringContext, is_answer_correct

FEEDBACK_DELAY = 5.0


def _active_players_count(room) -> int:
    return len([p for p in room.players.values() if not p.disconnected])


class GameFlow(ABC):
    def __init__(self, calculator: ScoreCalculator | None = None) -> None:
        self.calculator = calculator or ScoreCalculator()

    @abstractmethod
    def should_finish_round(self, room) -> bool:
        ...

    @abstractmethod
    def should_eliminate(self, room, player_id: str) -> bool:
        ...

    async def on_answer(
        self, room, room_id: str, player_id: str, background_tasks=None
    ) -> None:
        if self.should_finish_round(room):
            await self.finish_round(room_id, background_tasks)

    async def on_remove_player(
        self, room, room_id: str, background_tasks=None
    ) -> None:
        if (
            room.status == GameStatus.playing
            and room.feedback_until is None
            and self.should_finish_round(room)
        ):
            await self.finish_round(room_id, background_tasks)

    async def finish_round(self, room_id: str, background_tasks=None) -> None:
        room = await store._get_raw(room_id)
        if room is None or room.status != GameStatus.playing:
            return
        if room.feedback_until is not None:
            return
        if room.advance_task is not None:
            task = room.advance_task
            room.advance_task = None
            if task is not asyncio.current_task():
                task.cancel()
        for player in room.active_players:
            if player.player_id not in room.current_round_answers:
                room.current_round_answers[player.player_id] = RoundAnswer(
                    player_id=player.player_id,
                    selected_choices=[],
                    elapsed=room.timer,
                    timeout=True,
                )
                room.answered_players.add(player.player_id)
        results = await self._score_round(room)
        qstate = room.get_question_state(room.current_question_id())
        correct_choices = qstate.correct_choices if qstate else []
        room.feedback_until = time.time() + FEEDBACK_DELAY

        await notify_backend(
            room_id,
            "question-finished",
            {
                "question_id": room.current_question_id(),
                "correct_choices": correct_choices,
                "results": results,
            },
        )
        logger.info(
            "round_finished",
            extra={
                "room_id": room_id,
                "question_id": room.current_question_id(),
                "results": len(results),
            },
        )
        if background_tasks is not None:
            background_tasks.add_task(self._advance_after_feedback, room_id)
        else:
            room.advance_task = asyncio.create_task(
                self._advance_after_feedback(room_id)
            )

    async def _score_round(self, room) -> list[dict[str, Any]]:
        qstate = room.get_question_state(room.current_question_id())
        if qstate is None:
            return []

        correct_players = {
            ra.player_id
            for ra in room.current_round_answers.values()
            if not ra.timeout
            and is_answer_correct(
                qstate.correct_choices,
                ra.selected_choices,
                question_type=qstate.question_type,
            )
        }

        results: list[dict[str, Any]] = []
        ordered_answers = sorted(
            room.current_round_answers.values(), key=lambda ra: ra.elapsed
        )
        for ra in ordered_answers:
            player = room.players.get(ra.player_id)
            if player is None:
                continue
            old_streak = player.streak
            if ra.timeout:
                correct = False
                points = 0
                bonus = 0
                new_streak = 0
                player.streak = 0
                time_spent = float(room.timer)
            else:
                correct = is_answer_correct(
                    qstate.correct_choices,
                    ra.selected_choices,
                    question_type=qstate.question_type,
                )
                time_spent = min(ra.elapsed, room.timer)
                if correct:
                    first_correct = ra.player_id == next(
                        iter(correct_players), None
                    )
                    alone_correct = len(correct_players) == 1
                    ctx = ScoringContext(
                        mode=room.mode,
                        player_count=_active_players_count(room),
                        difficulty=qstate.difficulty,
                        is_correct=True,
                        current_streak=old_streak,
                        first_correct=first_correct,
                        alone_correct=alone_correct,
                    )
                    sr = self.calculator.calculate(ctx)
                    points = sr.total
                    bonus = sr.bonus
                    new_streak = sr.new_streak
                    player.streak = new_streak
                else:
                    points = 0
                    bonus = 0
                    new_streak = old_streak
                    player.streak = 0

            player.score += points
            player.cumulative_time += time_spent

            await store.record_answer(
                room.id,
                AnswerRecord(
                    player_id=ra.player_id,
                    question_id=qstate.id,
                    correct=correct,
                    time_spent=time_spent,
                ),
            )

            results.append(
                {
                    "player_id": ra.player_id,
                    "nickname": player.nickname,
                    "correct": correct,
                    "points": points,
                    "bonus": bonus,
                    "streak": player.streak if correct else 0,
                    "cumulative_time": player.cumulative_time,
                }
            )

        return results

    async def _advance_after_feedback(self, room_id: str) -> None:
        await asyncio.sleep(FEEDBACK_DELAY)
        await self.advance_question(room_id)

    async def advance_question(self, room_id: str) -> None:
        room = await store._get_raw(room_id)
        if room is None or room.status != GameStatus.playing:
            return

        room.current_question_index += 1
        room.answered_players.clear()
        room.current_round_answers.clear()
        room.feedback_until = None
        if room.advance_task is not None:
            task = room.advance_task
            room.advance_task = None
            if task is not asyncio.current_task():
                task.cancel()

        for player in room.players.values():
            player.current_question_index = room.current_question_index

        if room.current_question_index >= len(room.shuffled_question_ids):
            for player in room.players.values():
                player.finished = True
            room.status = GameStatus.finished
            await self._send_results(room_id)

            await notify_backend(room_id, "game-finished", {})
            logger.info("game_finished", extra={"room_id": room_id})
            return

        now = time.time()
        room.question_started_at = now
        room.question_deadline = now + room.timer
        room.advance_task = asyncio.create_task(
            self._deadline_task(room_id, room.question_deadline)
        )
        await notify_backend(
            room_id,
            "next-question",
            {"question_index": room.current_question_index},
        )
        logger.info(
            "next_question",
            extra={"room_id": room_id, "index": room.current_question_index},
        )

    async def _deadline_task(self, room_id: str, deadline: float) -> None:
        sleep_seconds = deadline - time.time()
        if sleep_seconds > 0:
            await asyncio.sleep(sleep_seconds)
        room = await store._get_raw(room_id)
        if room is None or room.status != GameStatus.playing:
            return
        if room.feedback_until is not None:
            return
        await self.finish_round(room_id)

    async def _send_results(self, room_id: str, max_retries: int = 3) -> None:
        room = await store._get_raw(room_id)
        if room is None:
            return

        sorted_players = sorted(
            room.players.values(),
            key=lambda p: (-p.score, p.cumulative_time),
        )

        from src.schemas import PlayerAnswer, ResultsCallback, ScoreboardEntry

        scores = [
            ScoreboardEntry(
                player_id=p.player_id,
                nickname=p.nickname,
                score=p.score,
                streak=p.streak,
                cumulative_time=p.cumulative_time,
            )
            for p in sorted_players
        ]

        answers = [
            PlayerAnswer(
                player_id=a.player_id,
                question_id=a.question_id,
                correct=a.correct,
                time_spent=a.time_spent,
            )
            for a in room.answers
        ]

        payload = ResultsCallback(scores=scores, answers=answers)
        await notify_backend(room_id, "results", payload.model_dump())


class ClassicFlow(GameFlow):
    def should_finish_round(self, room) -> bool:
        return room.all_active_answered()

    def should_eliminate(self, room, player_id: str) -> bool:
        return False
